modules/drivers/druid-jdbc/src/metabase/driver/druid_jdbc.clj

```diff
@@ -0,0 +1,183 @@
+(ns metabase.driver.druid-jdbc
+  (:require
+   [cheshire.core :as json]
+   [clj-http.client :as http]
+   [clojure.string :as str]
+   [clojure.walk :as walk]
+   [java-time.api :as t]
+   [metabase.driver :as driver]
+   [metabase.driver.common :as driver.common]
+   [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
+   [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
+   [metabase.driver.sql-jdbc.sync :as sql-jdbc.sync]
+   [metabase.driver.sql.query-processor :as sql.qp]
+   [metabase.driver.sql.query-processor.util :as sql.qp.u]
+   [metabase.driver.sql.util.unprepare :as unprepare]
+   [metabase.lib.field :as lib.field]
+   [metabase.lib.metadata :as lib.metadata]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.query-processor.util.add-alias-info :as add]
+   [metabase.util.honey-sql-2 :as h2x]
+   [metabase.util.log :as log])
+  (:import
+   (java.sql ResultSet Types)
+   (java.time ZonedDateTime)))
+
+(set! *warn-on-reflection* true)
+
+(driver/register! :druid-jdbc :parent :sql-jdbc)
+
+(doseq [[feature supported?] {:set-timezone            true
+                              :expression-aggregations true}]
+  (defmethod driver/database-supports? [:druid-jdbc feature] [_driver _feature _db] supported?))
+
+(defmethod sql-jdbc.conn/connection-details->spec :druid-jdbc
+  [_driver {:keys [host port] :as _db-details}]
+  (merge {:classname   "org.apache.calcite.avatica.remote.Driver"
+          :subprotocol "avatica:remote"
+          :subname     (str "url=" host ":" port "/druid/v2/sql/avatica/;transparent_reconnection=true")}
+         (when (some? (driver/report-timezone))
+           {:sqlTimeZone (driver/report-timezone)})))
```

Comment:
Can we set session timezone in `do-with-connection-with-options`? e.g. something like

```clj
(defmethod sql-jdbc.execute/do-with-connection-with-options :druid-jdbc
  [driver db-or-id-or-spec {:keys [session-timezone], :as options} f]
  (let [parent-method (get-method sql-jdbc.execute/do-with-connection-with-options :sql-jdbc)]
    (parent-method
     driver
     db-or-id-or-spec
     options
     (fn [^java.sql.Connection conn]
       (when session-timezone
         (with-open [stmt (.createStatement conn)]
           (.execute stmt (format "ALTER SESSION SET sqlTimeZone = '%s';" session-timezone))
       (f conn)))))
```

The main reason we would want to do that instead of as a parameter in the connection details spec is it really messes up the tests when we do it this way. Basically every time report timezone changes we will need to destroy the connection pool, which means we can't parallelize any tests that change report-timezone. I really want to get to a world where `mt/with-temporary-setting-values` is allowed inside `^:parallel` tests, and the one thing preventing that is that we have to destroy connection pools when report timezone changes. I think Snowflake is the only driver where we need to do this but I'm working on fixing that.

Reply:
To the best of my knowledge this is the only way how to set timezone for the Druid JDBC ([docs](https://druid.apache.org/docs/latest/querying/sql-query-context#example-using-jdbc)). Searching the docs yielded no clues for `ALTER` or anything similar.

On the other hand, Druid recommends using UTC not only in Druid instances but "for all data infrastructure" ([docs](https://druid.apache.org/docs/latest/operations/basic-cluster-tuning#use-utc-timezone)). Alternative would be disabling `:set-timezone` feature, but looking through our [past issues](https://github.com/metabase/metabase/issues?q=is%3Aissue+in%3Atitle+druid+timezone+) there are some users requiring that, hence that's probably not a great idea.

Reply:
Hmm... https://druid.apache.org/docs/latest/querying/sql-query-context/ seems to suggest that there is a way to set them using the REST API, I wonder if there's not some way we can set the query context for a SQL query as well, maybe by digging in to the Avatica internals or something. There is at least `AvaticaConnection.getTimeZone()` which would let us see what the current timezone is, but I can't see any easy way to change it for an existing connection.

Let's get your stuff merged-in as is and we can either disable parallel tests for Druid or dig into this further. We could probably make the connection pool stuff work without completely destroying connection pools when settings change, maybe we can just clear out the unused connections and then of course any new connections you fetch will have the correct settings